docs/source-2.0/glossary.rst

```diff
@@ -0,0 +1,152 @@
+        built on top of the :term:`Smithy reference implementation` that are
+        used to implement Smithy code generators. codegen-core contains
+        libraries for writing code, managing dependencies, managing imports,
+        converting Smithy shapes to :term:`symbols` of a :term:`target environment`,
+        :term:`reserved words` handling, etc.
+
+    Gradle
+        `Gradle <https://gradle.org/>`__ is a build tool for Java, Kotlin, and
+        other languages. Gradle is typically the build system used to develop
+        Smithy code generators. The Smithy team
+        `maintains a Gradle plugin <https://github.com/awslabs/smithy-gradle-plugin>`__ for
+        running Smithy-Build via Gradle.
+
+    Integrations
+        Integrations are code generator plugins. Integrations are defined by
+        each code generator. They can be used to preprocess the model, modify
+        the SymbolProvider used during code generation, add dependencies for
+        the target environment, generate additional files, register protocol
+        code generators, add configuration options to clients, etc.
+
+        .. seealso:: :doc:`guides/codegen/making-codegen-pluggable`
+
+    Java Service Provider Interface
+    SPI
+        A feature for loading and discovering implemenations of a Java interface.
+        SPI is used throughout the :term:`Smithy reference implementation` as
+        a plugin system. See the `Java documentation <https://docs.oracle.com/javase/tutorial/sound/SPI-intro.html>`__
+        for more information.
+
+    Knowledge index
+        Abstractions provided in the :term:`Smithy reference implementation`
+        that extract information from the metamodel in a more accessible way.
+        For example, the `HttpBindingIndex`_ makes it easier to codegen HTTP
+        bindings, and the `NullableIndex`_ hides the details of determining
+        if a member is optional or always present.
+
+    Projection
+        A specific view of a Smithy model that has added, removed, or
+        transformed model components.
+
+        .. seealso:: :doc:`guides/building-models/build-config`
+
+    Reserved words
+        Identifiers and words that cannot be used in a
+        :term:`target environment`. Reserved words can be contextual or global
+        to the target language (for example, a word might only reserved when
+        used as a structure property but not when used as the name of a
+        shape). Code generators are expected to automatically translate
+        reserved words into an identifier that is safe to use in the
+        target environment.
+
+    Runtime libraries
+        The libraries used at runtime in a :term:`target environment`.
+        For example, HTTP clients, type implementations like
+        big decimal, etc.
+
+    Semantic model
+        The Smithy semantic model is an in-memory representation the shapes,
```

Comment:
`shapes`, `traits`, and `Smithy reference implementation` should be `:term:` links